docs/pages/keyconcepts/state/settingstate.rst
.. _settingstate:


Setting State when Performing a Match
=====================================

A response includes two fields, `transitionsState` and `removesState` which alter Hoverflies internal state during a match:

.. code:: json

    "request": {
        "path": [
            {
                "matcher": "exact",
                "value": "/pay"
            }
        ]
    },
    "response": {
        "status": 200,
        "body": "eggs and large bacon",
        "transitionsState" : {
            "payment-flow" : "complete"
        },
        "removesState" : [
            "basket"
        ]
    }

In the above case, the following changes to Hoverflies internal state would be made on a match:

+----------------------------------+------------------------+----------------------------------------------------+
| Current State of Hoverfly        | New State of Hoverfly? | reason                                             |
+==================================+========================+====================================================+
| payment-flow=pending,basket=full | payment-flow=complete  | Payment value transitions, basket deleted by key   |
+----------------------------------+------------------------+----------------------------------------------------+
| basket=full                      | payment-flow=complete  | Payment value created, basket deleted by key       |
+----------------------------------+------------------------+----------------------------------------------------+
|                                  | payment-flow=complete  | Payment value created, basket already absent       |
+----------------------------------+------------------------+----------------------------------------------------+


docs/reference/flags_environment_variables.md
# Flags and environment variables

Hoverfly can be configured using flags on startup, or using environment variables.

## Admin UI Authentication

    -no-auth

Disable authentication, currently it is enabled by default.

    -add

Add a new user.

    -username <string>

Username for new user.

    -password <string>

Password for new user.

    -admin <bool>

Supply '-admin false' to make this a non-admin user (defaults to 'true').

For example:

    ./hoverfly -add -username hfadmin -password hfpass -admin false   	

This creates a new non-admin user with the username 'hfadmin' and the password 'hfpass'.

## Port selection

    -ap <string>

Sets the Admin UI port.

    -pp <string>

Sets the proxy port. For example:

    ./hoverfly -ap 1234 -pp 4567         

This starts Hoverfly with the Admin UI on port 1234 and the proxy on 4567.    	

## Mode selection, import & middleware

By default, Hoverfly starts in "virtualize" mode.

    -capture

Start Hoverfly in capture mode - transparently intercepts and saves requests/response.

    -synthesize

Start Hoverfly in synthesize mode. Middleware is required to generate responses for incoming requests.

    -modify

Start Hoverfly in modify mode. Middleware is required. Middleware is applied to both outgoing and incoming HTTP traffic.

    -middleware <string>

Set proxy to use middleware. Supply the path to the middleware script.

For example:

    ./hoverfly -synthesize -middleware "scripts/gen_response.py"

This starts Hoverfly in synthesize mode with a middleware script that generates responses on the fly.

    -import <string>

Import a virtual service JSON file from file system or URL. For example:

    ./hoverfly -import http://mypage.com/service_x.json

    ./hoverfly -import path/to/my/service_x.json      

This starts Hoverfly in virtualize mode (default mode), and imports a virtual service JSON file from either a URL or the local filesystem.

Note: currently, service JSON can only be exported from Hoverfly via the API.  

## Destination URL

    -dest <string>

Specify which hosts to process. For example:

    ./hoverfly -dest fooservice.org -dest barservice.org -dest catservice.org

This will start Hoverfly in virtualize mode, and only virtualize requests that are sent to fooservice.org, barservice.org and catservice.org. Requests to all other hosts will pass through.

    -destination <string>

Specify which URI to catch using regluar expression. (Defaults to ".").   

## Persistence

    -db <string>

Persistent storage to use. By default, Hoverfly uses BoltDB to store data in a file on disk. Specify 'memory'
to disable this and use in-memory cache only. 
Note: 'memory' should only be used for small test cases, if your machine runs out of RAM while Hoverfly is in 'memory' mode - 
bad things can happen.

    -db-path <string>

Path to BoltDB data file. By default, a "requests.db" file will be created in the directory from which Hoverfly is executed. 
Supply a custom path with filename to use a different file or location. The file will be created if it doesn't exist.

For example:

	 ./hoverfly -db-path new_name.db

## TLS & Certificate management

    -generate-ca-cert

Hoverfly will generate private and public keys in the current directory.

    -cert-name <string>

Certificate name (defaults to "hoverfly.proxy")

    -cert-org <string>

Organization name for new certificate (defaults to "Hoverfly Authority"). For example:

    ./hoverfly -generate-ca-cert -cert-name my_certificate -cert-org my_organization

This will create a certificate with the name "my_certificate" for the organization "my_organization" in the current directory.

    -cert <string>

Path to the certificate file to use.

    -key <string>

Path to the key file to use.  

    -tls-verification=<string>

Turn on/off TLS verification for outgoing requests (Hoverfly will not try to verify certificates) - defaults to true.	      

## Logging & metrics

    -v

Verbose mode. Logs every proxy and admin UI/API request to stdout.

    -metrics

Logs metrics to stdout every 5 seconds.


## Misc

    -dev

Supply -dev flag to serve Admin UI static files directly from ./static/admin/dist instead from statik binary. Useful when 
developing UI.

# Environment variables

You can configure Hoverfly through environment variables, this is a standard approach when application is running in a 
Docker container or hosted via platform. 

## Admin UI authentication

    HoverflyAuthDisabled  
    
Hoverfly authentication setting. Defaults to false. Set it to 'true' to disable authentication.
    
For example:
    
    export HoverflyAuthDisabled="true"
    
    HoverflySecret
    
Secret used to generate authentication tokens. If this variable is not set - Hoverfly generates one during startup.
Note: if you leave this value unset - you will have to authenticate in the admin UI after each Hoverfly restart.

    HoverflyTokenExpiration
    
Token expiration time in seconds. Defaults to one day (24 * 60 * 60).

### Setting admin user through env variables

You can provide credentials for initial user by setting these variables:

    HoverflyAdmin
    
Admin user name.

    HoverflyAdminPass
    
Admin password.
Note: if you do not set initial user through environment variables with authentication enabled - Hoverfly will ask you to input
username and password for the first user during startup. This could result in a 'stuck' container.

## Port selection 

    AdminPort
    
Admin UI port, defaults to 8500.
    
    ProxyPort
    
Proxy port, defaults to 8888.
    
## Persistence
 
    HoverflyDB

Path to BoltDB data file. By default, a "requests.db" file will be created in the directory from which Hoverfly is executed. 
Supply a custom path with filename to use a different file or location. The file will be created if it doesn't exist.

## TLS

    HoverflyTlsVerification
    
TLS verification. Since Hoverfly is making requests on behalf of it's users - it has to either trust remote servers or not.
This setting defaults to 'true' which means that untrusted hosts won't be accepted if request is done via HTTPs. You can turn it
off by providing "false" value to this environment variable, for example:

    export HoverflyTlsVerification="false"
    
    
    

core/settings.go
package hoverfly

import (
	"github.com/SpectoLabs/hoverfly/v2/core/cors"
	"os"
	"strconv"
	"sync"

	"strings"

	"github.com/SpectoLabs/hoverfly/v2/core/middleware"
	log "github.com/sirupsen/logrus"
)

// Configuration - initial structure of configuration
type Configuration struct {
	AdminPort    string
	ProxyPort    string
	ListenOnHost string
	Mode         string
	Destination  string
	Middleware   middleware.Middleware
	DatabasePath string
	Webserver    bool

	TLSVerification bool

	UpstreamProxy string
	PACFile       []byte

	Verbose bool

	DisableCache bool
	CacheSize    int

	SecretKey          []byte
	JWTExpirationDelta int
	AuthEnabled        bool

	ProxyAuthorizationHeader string

	PlainHttpTunneling bool
	CORS               cors.Configs

	NoImportCheck bool

	ClientAuthenticationDestination string
	ClientAuthenticationClientCert  string
	ClientAuthenticationClientKey   string
	ClientAuthenticationCACert      string

	ResponsesBodyFilesPath           string
	ResponsesBodyFilesAllowedOrigins []string

	ProxyControlWG sync.WaitGroup

	mu sync.Mutex
}

// SetMode - provides safe way to set new mode
func (c *Configuration) SetMode(mode string) {
	c.mu.Lock()
	c.Mode = mode
	c.mu.Unlock()
}

func (c *Configuration) SetUpstreamProxy(upstreamProxy string) {
	if !strings.HasPrefix(upstreamProxy, "http://") && !strings.HasPrefix(upstreamProxy, "https://") {
		upstreamProxy = "http://" + upstreamProxy
	}
	c.UpstreamProxy = upstreamProxy
}

// GetMode - provides safe way to get current mode
func (c *Configuration) GetMode() string {
	c.mu.Lock()
	mode := c.Mode
	c.mu.Unlock()
	return mode
}

// DefaultPort - default proxy port
const DefaultPort = "8500"

// DefaultAdminPort - default admin interface port
const DefaultAdminPort = "8888"

const DefaultListenOnHost = "127.0.0.1"

// DefaultDatabasePath - default database name that will be created
// or used by Hoverfly
const DefaultDatabasePath = "requests.db"

// DefaultJWTExpirationDelta - default token expiration if environment variable is no provided
const DefaultJWTExpirationDelta = 1 * 24 * 60 * 60

// Environment variables
const (
	// TODO Should use naming convention for environment variables
	HoverflyAuthEnabledEV     = "HoverflyAuthEnabled"
	HoverflySecretEV          = "HoverflySecret"
	HoverflyTokenExpirationEV = "HoverflyTokenExpiration"

	HoverflyAdminPortEV = "AdminPort"
	HoverflyProxyPortEV = "ProxyPort"

	HoverflyDBEV         = "HoverflyDB"
	HoverflyMiddlewareEV = "HoverflyMiddleware"

	HoverflyTLSVerification = "HoverflyTlsVerification"

	HoverflyAdminUsernameEV = "HoverflyAdmin"
	HoverflyAdminPasswordEV = "HoverflyAdminPass"

	HoverflyImportRecordsEV = "HoverflyImport"

	HoverflyUpstreamProxyPortEV = "UpstreamProxy"
	HoverflySkipImportCheckEV   = "SKIP_IMPORT_CHECK"
)

// InitSettings gets and returns initial configuration from env
// variables or sets defaults
func InitSettings() *Configuration {

	var appConfig Configuration
	// getting default admin interface port
	if os.Getenv(HoverflyAdminPortEV) != "" {
		appConfig.AdminPort = os.Getenv(HoverflyAdminPortEV)
	} else {
		appConfig.AdminPort = DefaultAdminPort
	}

	// getting proxy port
	if os.Getenv(HoverflyProxyPortEV) != "" {
		appConfig.ProxyPort = os.Getenv(HoverflyProxyPortEV)
	} else {
		appConfig.ProxyPort = DefaultPort
	}

	appConfig.ListenOnHost = DefaultListenOnHost

	// getting external proxy
	if os.Getenv(HoverflyUpstreamProxyPortEV) != "" {
		appConfig.UpstreamProxy = os.Getenv(HoverflyUpstreamProxyPortEV)
	} else {
		appConfig.UpstreamProxy = ""
	}

	databasePath := os.Getenv(HoverflyDBEV)
	if databasePath == "" {
		appConfig.DatabasePath = DefaultDatabasePath
	}

	appConfig.Webserver = false

	if os.Getenv(HoverflySecretEV) != "" {
		appConfig.SecretKey = []byte(os.Getenv(HoverflySecretEV))
	} else {
		appConfig.SecretKey = getRandomName(10)
	}

	if os.Getenv(HoverflyTokenExpirationEV) != "" {

		exp, err := strconv.Atoi(os.Getenv(HoverflyTokenExpirationEV))
		if err != nil {
			log.WithFields(log.Fields{
				"error":                   err.Error(),
				"HoverflyTokenExpiration": os.Getenv(HoverflyTokenExpirationEV),
			}).Fatal("failed to get token exipration delta, using default value")
			exp = DefaultJWTExpirationDelta
		}
		appConfig.JWTExpirationDelta = exp

	} else {
		appConfig.JWTExpirationDelta = DefaultJWTExpirationDelta
	}

	if os.Getenv(HoverflyAuthEnabledEV) == "true" {
		appConfig.AuthEnabled = true
	} else {
		appConfig.AuthEnabled = false
	}

	// middleware configuration
	newMiddleware, _ := middleware.ConvertToNewMiddleware(os.Getenv(HoverflyMiddlewareEV))

	appConfig.Middleware = *newMiddleware

	if os.Getenv(HoverflyTLSVerification) == "false" {
		appConfig.TLSVerification = false
	} else {
		appConfig.TLSVerification = true
	}

	if os.Getenv(HoverflySkipImportCheckEV) == "true" {
		appConfig.NoImportCheck = true
	} else {
		appConfig.NoImportCheck = false
	}

	appConfig.Mode = "simulate"

	appConfig.ProxyAuthorizationHeader = "Proxy-Authorization"

	appConfig.CacheSize = 1000

	return &appConfig
}
